### PR TITLE
ENH: improve dataprobe magnifier borders

### PR DESCRIPTION
**Pull request by fbordignon**
The DataProbe magnifier currently has an odd behavior when the mouse approach the corners of the slice views:
image

This PR uses a 2D canvas with a fixed size and draws the crop imagedata on it at the right position.

### CODE REVIEW

**Review by pieper — DISMISSED** [review]
Looks good to me.  Anyone else want to have a look?

**Review by lassoan — COMMENTED** [review]
It looks OK, I've just added a few minor comments.

Can you tell how you use this feature? I'm just curious because I find it completely useless:
- the mouse pointer position is often not the target point position (e.g., when you reposition a markup point then the red crosshair in the zoomed view does not show the actual markup point position)
- you don't see any annotations (slice intersection, markups, segmentations, etc.)
- precise placement is hard because the mouse is not slowed down

All these issues are solved very well by built-in operating system accessibility features (see for example the Magnifier tool on Windows - hit <kbd>Win</kbd>-<kbd>+</kbd> to show it and <kbd>Win</kbd>-<kbd>Esc</kbd> to hide).

If you don't want to use operating system features then an equivalent solution to all the issues is to zoom in the view - maybe we would just need a keyboard shortcut to zoom in/out at the current mouse position by a factor of 4x?

If temporarily zooming in is not acceptable then we could think about adding a real slice view to the dataprobe, which would mimic the settings in the current slice view except it would zoom it.

**Comment by fbordignon** [comment]
Well, this feature was already present since the first time I remember using Slicer. I did not add this feature, I've simply improved the behavior when the mouse reaches the borders.

> It looks OK, I've just added a few minor comments.
> 
> Can you tell how you use this feature? I'm just curious because I find it completely useless:

It is useful because you don't need to look at your mouse and then look at the DataProbe to inspect the ijk, ras and scalar/rgb values. Sometimes the user would position the mouse and by the time you look at the dataprobe you are not certain anymore if the pointer is positioned at the right position, i.e. the mouse moved a bit and you need to look back and forth between the slice view and the dataprobe to see where the mouse is and what the values are. Having a small zoomed section of the view that shows the zoomed image close to the DataProbe values is good, in my opinion.

> * the mouse pointer position is often not the target point position (e.g., when you reposition a markup point then the red crosshair in the zoomed view does not show the actual markup point position)
> * you don't see any annotations (slice intersection, markups, segmentations, etc.)
> * precise placement is hard because the mouse is not slowed down

You are right! Maybe connecting to the final output of the sliceview can solve this, I did not look into the input of the method, but it is getting the output of the sliceview Blend, so it considers only the background and foreground.

> All these issues are solved very well by built-in operating system accessibility features (see for example the Magnifier tool on Windows - hit Win-+ to show it and Win-Esc to hide).

The OS magnifier does not encompass the use case of checking the values of ijk, ras and scalar/rgb values while positioning the mouse without the need to shift your focus to the opposite side of the screen.

> If you don't want to use operating system features then an equivalent solution to all the issues is to zoom in the view - maybe we would just need a keyboard shortcut to zoom in/out at the current mouse position by a factor of 4x?
> 
> If temporarily zooming in is not acceptable then we could think about adding a real slice view to the dataprobe, which would mimic the settings in the current slice view except it would zoom it.

Could work, but the crosshair is nice to see where exactly you are positioning the mouse, maybe connecting to the final image of the SliceView can do the trick

**Comment by jamesobutler** [comment]
@fbordignon https://github.com/Slicer/Slicer/pull/5142 may be of interest to you if you use this feature.

File: Modules/Scripted/DataProbe/DataProbe.py [hunk 1]
```diff
@@ -350,14 +352,25 @@ def _roundInt(value):
...
     imageSize = _roundInt(minDim/imageZoom/2.0)
-    imin = max(0,xyzInt[0]-imageSize)
-    imax = min(dims[0]-1,  xyzInt[0]+imageSize)
-    jmin = max(0,xyzInt[1]-imageSize)
-    jmax = min(dims[1]-1,  xyzInt[1]+imageSize)
+    imin = xyzInt[0]-imageSize
+    imax = xyzInt[0]+imageSize
+    jmin = xyzInt[1]-imageSize
+    jmax = xyzInt[1]+imageSize
+    canvas = self.canvas
```
[the remark this answers is not in the document]
Comment (fbordignon) [reply Modules/Scripted/DataProbe/DataProbe.py:362]: ✔️

File: Modules/Scripted/DataProbe/DataProbe.py [hunk 2]
```diff
@@ -350,14 +352,25 @@ def _roundInt(value):
...
+    canvas.SetScalarType(producer.GetOutput().GetScalarType())
+    canvas.SetNumberOfScalarComponents(producer.GetOutput().GetNumberOfScalarComponents())
+    canvas.SetExtent(imin, imax, jmin , jmax, 0 ,0)
+    canvas.FillBox(imin, imax, jmin , jmax)
+    canvas.Update()
     if (imin <= imax) and (jmin <= jmax):
       imageCrop.SetVOI(imin, imax, jmin, jmax, 0,0)
       imageCrop.Update()
-      vtkImage = imageCrop.GetOutput()
+      vtkImage2 = imageCrop.GetOutput()
```
[the remark this answers is not in the document]
Comment (fbordignon) [reply Modules/Scripted/DataProbe/DataProbe.py]: ✔️

File: Modules/Scripted/DataProbe/DataProbe.py [hunk 3]
```diff
@@ -6,6 +6,7 @@
 import DataProbeLib
 from slicer.ScriptedLoadableModule import *
 from slicer.util import TESTING_DATA_URL
+import numpy
```
[the remark this answers is not in the document]
Comment (fbordignon) [reply Modules/Scripted/DataProbe/DataProbe.py:9]: ✔️

**Comment by fbordignon** [comment]
> @fbordignon #5142 may be of interest to you if you use this feature.

Thanks, that is cool. But my main use for the zoom at the DataProbe is to probe data while looking at the image all in the same region of the screen. This way I don't need to shift focus from the data probe panel to the slice view to see where the mouse is pointing at and what are the data values, ijk and ras coordinates.

**Review by lassoan — CHANGES_REQUESTED** [review]
Thanks for the updates. I've tested this and the behavior looks good. Unfortunately, the application log gets filled with errors like these:

```
[ERROR][VTK] 14.03.2022 17:43:08 [vtkCompositeDataPipeline (0000026BE5949980)] (C:\D\S4R\VTK\Common\ExecutionModel\vtkStreamingDemandDrivenPipeline.cxx:878) - The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0000026BE01122A0) is 1213 1323 -36 74 0 0, which is outside the whole extent 0 1327 0 1091 0 0.
[ERROR][VTK] 14.03.2022 17:43:29 [vtkCompositeDataPipeline (0000026BE5949980)] (C:\D\S4R\VTK\Common\ExecutionModel\vtkStreamingDemandDrivenPipeline.cxx:878) - The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0000026BE01122A0) is -29 81 64 174 0 0, which is outside the whole extent 0 1327 0 1091 0 0.
[ERROR][VTK] 14.03.2022 17:44:37 [vtkCompositeDataPipeline (0000026BE5949980)] (C:\D\S4R\VTK\Common\ExecutionModel\vtkStreamingDemandDrivenPipeline.cxx:878) - The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0000026BE01122A0) is 467 577 -34 76 0 0, which is outside the whole extent 0 1327 0 1091 0 0.
[ERROR][VTK] 14.03.2022 17:44:37 [vtkCompositeDataPipeline (0000026BE5949980)] (C:\D\S4R\VTK\Common\ExecutionModel\vtkStreamingDemandDrivenPipeline.cxx:878) - The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0000026BE01122A0) is 473 583 1012 1122 0 0, which is outside the whole extent 0 1327 0 1091 0 0.
```

Could you please update the code to avoid these errors? Thank you!

**Comment by fbordignon** [comment]
Thanks for checking the error logs. I removed the truncated i/j mins and maxes to create the canvas with a fixed size and in practice it did not altered the behavior of the cropping but issued the abovementioned errors. I've brought the truncated values back and used it when pertinent.

**Review by lassoan — APPROVED** [review]
It looks good, thank you.

**Pull request merged**